### ISSUE TITLE
scipy.stats.binned_statistic regressed in v1.4.0

### ISSUE THREAD

**Issue by barentsen** [scipy/scipy#11365]
## Summary

PR #10664 recently changed the behavior of `scipy.stats.binned_statistic` to raise a `ValueError` whenever the data contains a non-finite number (e.g., `nan`, `inf`).

This new behavior is counter-intuitive because many other statistical methods do not raise exceptions in the presence of NaNs, e.g. `np.mean([np.nan])` returns `nan` rather than raising an exception.

This change was included in SciPy v1.4.0.  I propose that we undo #10664 as part of v1.4.2 for three reasons:
1. The change caused `binned_statistic` to start raising unnecessary exceptions in common use-cases that worked fine before (see examples below). 
2. It is common to have nans and infs present in scientific data, so the change will likely affect (crash) a lot of existing software (e.g. it broke a feature of the Lightkurve package; https://github.com/KeplerGO/lightkurve/pull/661).
3. I believe the change can be reverted without negative consequences.

## Examples
In SciPy v1.3.x and earlier, `binned_statistic` allowed NaN values to be present.  For example:
```python
>>> import scipy.stats, numpy as np
>>> x = [0.5, 0.5, 1.5, 1.5]
>>> values = [10, 20, np.nan, 40]
>>> scipy.stats.binned_statistic(x, values, statistic='mean', bins=(0, 1, 2)).statistic
array([15., nan])
```

This behavior was sensible because it is consistent with e.g. `np.mean`:
```python
>> np.mean([np.nan, 40])
nan
```

A user could choose to ignore the NaNs by providing a custom statistic function, e.g. `np.nanmean`:
```python
>>> scipy.stats.binned_statistic(x, values, statistic=np.nanmean, bins=(0, 1, 2)).statistic
array([15., 40.])
```

As of SciPy v1.4.0, the examples above raise this `ValueError`:
```python
>>> scipy.stats.binned_statistic(x, values, statistic=np.nanmean, bins=(0, 1, 2)).statistic
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gb/bin/anaconda/envs/sp14/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py", line 180, in binned_statistic
    [x], values, statistic, bins, range)
  File "/home/gb/bin/anaconda/envs/sp14/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py", line 519, in binned_statistic_dd
    raise ValueError('%r or %r contains non-finite values.' % (sample, values,))
ValueError: [[0.5, 0.5, 1.5, 1.5]] or [10, 20, nan, 40] contains non-finite values.
```

## Cause

The behavior changed because these two lines were added via #10664:
https://github.com/scipy/scipy/blob/729efb880058cf841a561c733910f2aefb0a9cf6/scipy/stats/_binned_statistic.py#L518-L519

(Unrelated note: there is a syntax error at the end of line 518; round brackets are missing after `any`.)

This change was made in response to issue #9010, where a user reported an exception triggered by the use of a statistic that did not deal gracefully with the presence of NaNs.  While I agree it would be worth improving the error message in such a situation, I believe that doing this by refusing NaNs altogether is a negative and unintended regression.

**Comment by rlucas7** [comment]
@barentsen Sorry to break the code in lightcurve, that is my PR (and sadly, my syntax error).
I can open a fix but I'm not sure going back to no warning and default propagation of `nan` and `inf` is the preferred approach. The error messaging is a bit confusing if it hits an error and is difficult to debug. 

Would a `RuntimeWarning` work in this case or is that still going to break the code in lightcurve?

If `binned_statistic_dd()` emits a runtime warning the code would still execute (as compard to now where an exception thrown) but `binned_statistic_dd()` would notify of what is happening and make it easier on the end user to debug.

**Comment by jakevdp** [comment]
I'm generally not in favor of warnings that may show up by default during supported uses of an API, and I think the motivating error in #9010 was clear enough as-is (the error came from a user-defined custom aggregate that explicitly does not support infinities.

I'd be in favor of reverting the changes in #10664 and restoring the previous behavior.

**Comment by tylerjereddy** [comment]
Keep me posted if I need to backport something. I think I'll wait for a few more `stats` regulars to chime in :)

**Comment by barentsen** [comment]
Having re-read #9010 more carefully (in particular the later comments in the thread), I see that there is a specific case where `binned_statistic` yields an obscure error if `x` contains a NaN and `bins` is an integer. This is probably what motivated  #10664. Example:

```python
>>> import scipy.stats, numpy as np
>>> x = [0.5, 0.5, 1.5, np.nan]
>>> values = [10, 20, 30, 40]
>>> scipy.stats.binned_statistic(x, values, bins=3)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0df7a9f2770e> in <module>
      2 x = [0.5, 0.5, 1.5, np.nan]
      3 values = [10, 20, 30, 40]
----> 4 scipy.stats.binned_statistic(x, values, bins=3)

~/bin/anaconda/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py in binned_statistic(x, values, statistic, bins, range)
    179 
    180     medians, edges, binnumbers = binned_statistic_dd(
--> 181         [x], values, statistic, bins, range)
    182 
    183     return BinnedStatisticResult(medians, edges[0], binnumbers)

~/bin/anaconda/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py in binned_statistic_dd(sample, values, statistic, bins, range, expand_binnumbers)
    540     for i in xrange(Ndim):
    541         # Find the rounding precision
--> 542         decimal = int(-np.log10(dedges[i].min())) + 6
    543         # Find which points are on the rightmost edge.
    544         on_edge = np.where(np.around(sample[:, i], decimal) ==

ValueError: cannot convert float NaN to integer
```

The issue appears to be that the `_bin_edges()` helper function is not robust against the presence of NaNs in `x`.  Unless we modify that, I would be fine with an exception being raised if `x` contains NaNs and bins is an integer.  The `binned_statistic` is already crashing in this scenario anyway.

**However** I am unable to reproduce an example where a NaN in `values` (rather than `x`) leads to an obscure crash caused by binned_statistic, so I don't think there is an argument for enforcing that `values` cannot have NaNs. 

In summary: one possible solution is to modify `binned_statistic_dd` to say this:

```python
if isinstance(bins, int) and not np.isfinite(x).all():
    raise ValueError(...)
```
instead of this:
```python
if not np.isfinite(values).all() or not np.isfinite(sample).all:
    raise ValueError(...)
```

**Comment by rlucas7** [comment]
> I'm generally not in favor of warnings that show up by design during supported uses of an API, 
@jakevdp Yes, agreed.

**Comment by rlucas7** [comment]
> Having re-read #9010 more carefully (in particular the later comments in the thread), I see that there is a specific case where `binned_statistic` yields an obscure error if `x` contains a NaN and `bins` is an integer. This is probably what motivated #10664. Example:
> 
> ```python
> >>> import scipy.stats, numpy as np
> >>> x = [0.5, 0.5, 1.5, np.nan]
> >>> values = [10, 20, 30, 40]
> >>> scipy.stats.binned_statistic(x, values, bins=3)
> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> <ipython-input-3-0df7a9f2770e> in <module>
>       2 x = [0.5, 0.5, 1.5, np.nan]
>       3 values = [10, 20, 30, 40]
> ----> 4 scipy.stats.binned_statistic(x, values, bins=3)
> 
> ~/bin/anaconda/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py in binned_statistic(x, values, statistic, bins, range)
>     179 
>     180     medians, edges, binnumbers = binned_statistic_dd(
> --> 181         [x], values, statistic, bins, range)
>     182 
>     183     return BinnedStatisticResult(medians, edges[0], binnumbers)
> 
> ~/bin/anaconda/lib/python3.7/site-packages/scipy/stats/_binned_statistic.py in binned_statistic_dd(sample, values, statistic, bins, range, expand_binnumbers)
>     540     for i in xrange(Ndim):
>     541         # Find the rounding precision
> --> 542         decimal = int(-np.log10(dedges[i].min())) + 6
>     543         # Find which points are on the rightmost edge.
>     544         on_edge = np.where(np.around(sample[:, i], decimal) ==
> 
> ValueError: cannot convert float NaN to integer
> ```
> 
> The issue appears to be that the `_bin_edges()` helper function is not robust against the presence of NaNs in `x`. Unless we modify that, I would be fine with an exception being raised if `x` contains NaNs and bins is an integer. The `binned_statistic` is already crashing in this scenario anyway.

This looks like the same as in the original issue and I also confirmed the error locally. 

> **However** I am unable to reproduce an example where a NaN in `values` (rather than `x`) leads to an obscure crash caused by binned_statistic, so I don't think there is an argument for enforcing that `values` cannot have NaNs.

I haven't had a chance to look at this part closely yet, I'll take a closer look over the upcoming weekend and comment. 

> In summary: one possible solution is to modify `binned_statistic_dd` to say this:
> 
> ```python
> if isinstance(bins, int) and not np.isfinite(x).all():
>     raise ValueError(...)
> ```
> 
> instead of this:
> 
> ```python
> if not np.isfinite(values).all() or not np.isfinite(sample).all:
>     raise ValueError(...)
> ```

I'll take a closer look this weekend, looks correct but I didn't investigate locally.

**Comment by rlucas7** [comment]
I had a little bit of time this evening to look into the error. I think the proposed changes: 
```
if isinstance(bins, int) and not np.isfinite(x).all():
    raise ValueError(...)
```
will work.
Although I'm not super keen on having the check and error raising inside the `_binned_statistic_dd()` method when the error should be thrown by `_bin_edges()`. 
However, throwing it in `_bin_edges()` is a private method (and new in 1.4.0 release) so most users won't know what it is/does.

**Comment by rlucas7** [comment]
> Keep me posted if I need to backport something. I think I'll wait for a few more `stats` regulars to chime in :)

@tylerjereddy I'm not sure if you want to have other `stats` folks weight on this issue. I'm in agreement on fix and just opened a PR that's linked w/changes. Labeled with backport candidate and needs-decision to match w/status of this issue.

**Comment by rgommers** [comment]
> However, throwing it in `_bin_edges()` is a private method (and new in 1.4.0 release) so most users won't know what it is/does.

I don't think that matters much, as long as the exception message is clear.

Agreed that this change was unintentional. Would be good to replace so the original corner case is still handled but the reported regression is fixed, and then backporting it to `1.4.2`.

**Comment by rlucas7** [comment]
> > However, throwing it in `_bin_edges()` is a private method (and new in 1.4.0 release) so most users won't know what it is/does.
> 
> I don't think that matters much, as long as the exception message is clear.

Fair point, I guess `_bin_edges()` is "private" method so shouldn't really matter to end user, I was thinking in terms of maintenance and development, e.g. understanding where the root cause resides.  
 
> Agreed that this change was unintentional. Would be good to replace so the original corner case is still handled but the reported regression is fixed, and then backporting it to `1.4.2`.

https://github.com/scipy/scipy/issues/9010#issuecomment-520680191

was the issue you mentioned that I misread, that should be fixed in: 
https://github.com/scipy/scipy/pull/11382/files

the test has a case with a `np.inf` but no `np.nan` converage. The numpy method `isfinite()` should catch the `np.nan` case as well but there isn't a test for that one. I'll add that to the PR-thanks.

**Comment by rlucas7** [comment]
Related (recently made) comment in the commit:
https://github.com/scipy/scipy/commit/c509c0fef73fba0b65b03514e1920b9347b2e633#commitcomment-37401479
